core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala

```diff
@@ -226,8 +243,11 @@ private[spark] class ExecutorAllocationManager(
    * If the cap on the number of executors is reached, give up and reset the
    * number of executors to add next round instead of continuing to double it.
    * Return the number actually requested.
+   *
+   * @param maxNumNeededExecutors the maximum number of executors all currently running or pending
+   *                              tasks could fill
    */
-  private def addExecutors(): Int = synchronized {
+  private def addExecutors(maxNumNeededExecutors: Int): Int = synchronized {
     // Do not request more executors if we have already reached the upper bound
     val numExistingExecutors = executorIds.size + numExecutorsPending
```

Comment:
i think there we should make  numExistingExecutors  exclude executorsPendingToRemove.size?